File: src/test/java/cz/vhromada/catalog/repository/ShowRepositoryIntegrationTest.java
package cz.vhromada.catalog.repository;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.SoftAssertions.assertSoftly;

import java.util.Collections;
import java.util.List;

import javax.persistence.EntityManager;

import cz.vhromada.catalog.CatalogTestConfiguration;
import cz.vhromada.catalog.domain.Season;
import cz.vhromada.catalog.domain.Show;
import cz.vhromada.catalog.utils.EpisodeUtils;
import cz.vhromada.catalog.utils.GenreUtils;
import cz.vhromada.catalog.utils.SeasonUtils;
import cz.vhromada.catalog.utils.ShowUtils;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.data.domain.Sort;
import org.springframework.test.annotation.Rollback;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.transaction.annotation.Transactional;

/**
 * A class represents integration test for class {@link ShowRepository}.
 *
 * @author Vladimir Hromada
 */
@ExtendWith(SpringExtension.class)
@ContextConfiguration(classes = CatalogTestConfiguration.class)
@Transactional
@Rollback
class ShowRepositoryIntegrationTest {

    /**
     * Instance of {@link EntityManager}
     */
    @Autowired
    @Qualifier("containerManagedEntityManager")
    private EntityManager entityManager;

    /**
     * Instance of {@link ShowRepository}
     */
    @Autowired
    private ShowRepository showRepository;

    /**
     * Test method for get shows.
     */
    @Test
    void getShows() {
        final List<Show> shows = showRepository.findAll(Sort.by("position", "id"));

        ShowUtils.assertShowsDeepEquals(ShowUtils.getShows(), shows);

        assertSoftly(softly -> {
            softly.assertThat(ShowUtils.getShowsCount(entityManager)).isEqualTo(ShowUtils.SHOWS_COUNT);
            softly.assertThat(SeasonUtils.getSeasonsCount(entityManager)).isEqualTo(SeasonUtils.SEASONS_COUNT);
            softly.assertThat(EpisodeUtils.getEpisodesCount(entityManager)).isEqualTo(EpisodeUtils.EPISODES_COUNT);
        });
    }

    /**
     * Test method for get show.
     */
    @Test
    void getShow() {
        for (int i = 1; i <= ShowUtils.SHOWS_COUNT; i++) {
            final Show show = showRepository.findById(i).orElse(null);

            ShowUtils.assertShowDeepEquals(ShowUtils.getShow(i), show);
        }

        assertThat(showRepository.findById(Integer.MAX_VALUE).isPresent()).isFalse();

        assertSoftly(softly -> {
            softly.assertThat(ShowUtils.getShowsCount(entityManager)).isEqualTo(ShowUtils.SHOWS_COUNT);
            softly.assertThat(SeasonUtils.getSeasonsCount(entityManager)).isEqualTo(SeasonUtils.SEASONS_COUNT);
            softly.assertThat(EpisodeUtils.getEpisodesCount(entityManager)).isEqualTo(EpisodeUtils.EPISODES_COUNT);
        });
    }

    /**
     * Test method for add show.
     */
    @Test
    void add() {
        final Show show = ShowUtils.newShowDomain(null);
        show.setPosition(ShowUtils.SHOWS_COUNT);
        show.setGenres(Collections.singletonList(GenreUtils.getGenre(entityManager, 1)));

        showRepository.save(show);

        assertThat(show.getId()).isEqualTo(ShowUtils.SHOWS_COUNT + 1);

        final Show addedShow = ShowUtils.getShow(entityManager, ShowUtils.SHOWS_COUNT + 1);
        final Show expectedAddedShow = ShowUtils.newShowDomain(null);
        expectedAddedShow.setId(ShowUtils.SHOWS_COUNT + 1);
        expectedAddedShow.setPosition(ShowUtils.SHOWS_COUNT);
        expectedAddedShow.setGenres(Collections.singletonList(GenreUtils.getGenreDomain(1)));
        ShowUtils.assertShowDeepEquals(expectedAddedShow, addedShow);

        assertSoftly(softly -> {
            softly.assertThat(ShowUtils.getShowsCount(entityManager)).isEqualTo(ShowUtils.SHOWS_COUNT + 1);
            softly.assertThat(SeasonUtils.getSeasonsCount(entityManager)).isEqualTo(SeasonUtils.SEASONS_COUNT);
            softly.assertThat(EpisodeUtils.getEpisodesCount(entityManager)).isEqualTo(EpisodeUtils.EPISODES_COUNT);
        });
    }

    /**
     * Test method for update show with updated data.
     */
    @Test
    void update_Data() {
        final Show show = ShowUtils.updateShow(entityManager, 1);

        showRepository.save(show);

        final Show updatedShow = ShowUtils.getShow(entityManager, 1);
        final Show expectedUpdatedShow = ShowUtils.getShow(1);
        ShowUtils.updateShow(expectedUpdatedShow);
        expectedUpdatedShow.setPosition(ShowUtils.POSITION);
        ShowUtils.assertShowDeepEquals(expectedUpdatedShow, updatedShow);

        assertSoftly(softly -> {
            softly.assertThat(ShowUtils.getShowsCount(entityManager)).isEqualTo(ShowUtils.SHOWS_COUNT);
            softly.assertThat(SeasonUtils.getSeasonsCount(entityManager)).isEqualTo(SeasonUtils.SEASONS_COUNT);
            softly.assertThat(EpisodeUtils.getEpisodesCount(entityManager)).isEqualTo(EpisodeUtils.EPISODES_COUNT);
        });
    }

    /**
     * Test method for update show with added season.
     */
    @Test
    void update_AddedSeason() {
        final Season season = SeasonUtils.newSeasonDomain(null);
        season.setPosition(SeasonUtils.SEASONS_COUNT);
        entityManager.persist(season);

        final Show show = ShowUtils.getShow(entityManager, 1);
        show.getSeasons().add(season);

        showRepository.save(show);

        final Show updatedShow = ShowUtils.getShow(entityManager, 1);
        final Season expectedSeason = SeasonUtils.newSeasonDomain(null);
        expectedSeason.setId(SeasonUtils.SEASONS_COUNT + 1);
        expectedSeason.setPosition(SeasonUtils.SEASONS_COUNT);
        final Show expectedUpdatedShow = ShowUtils.getShow(1);
        expectedUpdatedShow.getSeasons().add(expectedSeason);
        ShowUtils.assertShowDeepEquals(expectedUpdatedShow, updatedShow);

        assertSoftly(softly -> {
            softly.assertThat(ShowUtils.getShowsCount(entityManager)).isEqualTo(ShowUtils.SHOWS_COUNT);
            softly.assertThat(SeasonUtils.getSeasonsCount(entityManager)).isEqualTo(SeasonUtils.SEASONS_COUNT + 1);
            softly.assertThat(EpisodeUtils.getEpisodesCount(entityManager)).isEqualTo(EpisodeUtils.EPISODES_COUNT);
        });
    }

    /**
     * Test method for remove show.
     */
    @Test
    void remove() {
        final int seasonsCount = ShowUtils.getShow(1).getSeasons().size();
        final int episodesCount = seasonsCount * EpisodeUtils.EPISODES_PER_SEASON_COUNT;

        showRepository.delete(ShowUtils.getShow(entityManager, 1));

        assertThat(ShowUtils.getShow(entityManager, 1)).isNull();

        assertSoftly(softly -> {
            softly.assertThat(ShowUtils.getShowsCount(entityManager)).isEqualTo(ShowUtils.SHOWS_COUNT - 1);
            softly.assertThat(SeasonUtils.getSeasonsCount(entityManager)).isEqualTo(SeasonUtils.SEASONS_COUNT - seasonsCount);
            softly.assertThat(EpisodeUtils.getEpisodesCount(entityManager)).isEqualTo(EpisodeUtils.EPISODES_COUNT - episodesCount);
        });
    }

    /**
     * Test method for remove all shows.
     */
    @Test
    void removeAll() {
        showRepository.deleteAll();

        assertSoftly(softly -> {
            softly.assertThat(ShowUtils.getShowsCount(entityManager)).isEqualTo(0);
            softly.assertThat(SeasonUtils.getSeasonsCount(entityManager)).isEqualTo(0);
            softly.assertThat(EpisodeUtils.getEpisodesCount(entityManager)).isEqualTo(0);
        });
    }

}

File: catalog-core/src/test/kotlin/com/github/vhromada/catalog/repository/ShowRepositoryIntegrationTest.kt
package com.github.vhromada.catalog.repository

import com.github.vhromada.catalog.CatalogTestConfiguration
import com.github.vhromada.catalog.utils.AuditUtils
import com.github.vhromada.catalog.utils.EpisodeUtils
import com.github.vhromada.catalog.utils.GenreUtils
import com.github.vhromada.catalog.utils.SeasonUtils
import com.github.vhromada.catalog.utils.ShowUtils
import com.github.vhromada.catalog.utils.TestConstants
import com.github.vhromada.catalog.utils.fillAudit
import com.github.vhromada.catalog.utils.updated
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.SoftAssertions.assertSoftly
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.annotation.Rollback
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.transaction.annotation.Transactional
import javax.persistence.EntityManager
import javax.persistence.PersistenceContext

/**
 * A class represents integration test for class [ShowRepository].
 *
 * @author Vladimir Hromada
 */
@ExtendWith(SpringExtension::class)
@ContextConfiguration(classes = [CatalogTestConfiguration::class])
@Transactional
@Rollback
class ShowRepositoryIntegrationTest {

    /**
     * Instance of [EntityManager]
     */
    @PersistenceContext
    private lateinit var entityManager: EntityManager

    /**
     * Instance of [ShowRepository]
     */
    @Autowired
    private lateinit var repository: ShowRepository

    /**
     * Test method for get shows.
     */
    @Test
    fun getShows() {
        val shows = repository.findAll()

        ShowUtils.assertDomainShowsDeepEquals(expected = ShowUtils.getShows(), actual = shows)

        assertSoftly {
            it.assertThat(ShowUtils.getShowsCount(entityManager = entityManager)).isEqualTo(ShowUtils.SHOWS_COUNT)
            it.assertThat(SeasonUtils.getSeasonsCount(entityManager = entityManager)).isEqualTo(SeasonUtils.SEASONS_COUNT)
            it.assertThat(EpisodeUtils.getEpisodesCount(entityManager = entityManager)).isEqualTo(EpisodeUtils.EPISODES_COUNT)
        }
    }

    /**
     * Test method for get show.
     */
    @Test
    fun getShow() {
        for (i in 1..ShowUtils.SHOWS_COUNT) {
            val show = repository.findById(i).orElse(null)

            ShowUtils.assertShowDeepEquals(expected = ShowUtils.getShowDomain(index = i), actual = show)
        }

        assertThat(repository.findById(Int.MAX_VALUE)).isNotPresent

        assertSoftly {
            it.assertThat(ShowUtils.getShowsCount(entityManager = entityManager)).isEqualTo(ShowUtils.SHOWS_COUNT)
            it.assertThat(SeasonUtils.getSeasonsCount(entityManager = entityManager)).isEqualTo(SeasonUtils.SEASONS_COUNT)
            it.assertThat(EpisodeUtils.getEpisodesCount(entityManager = entityManager)).isEqualTo(EpisodeUtils.EPISODES_COUNT)
        }
    }

    /**
     * Test method for add show.
     */
    @Test
    fun add() {
        val show = ShowUtils.newShowDomain(id = null)
            .copy(position = ShowUtils.SHOWS_COUNT, genres = listOf(GenreUtils.getGenre(entityManager = entityManager, id = 1)!!))
        val expectedShow = ShowUtils.newShowDomain(id = ShowUtils.SHOWS_COUNT + 1)
            .copy(picture = null, genres = listOf(GenreUtils.getGenre(entityManager = entityManager, id = 1)!!))
            .fillAudit(audit = AuditUtils.newAudit())

        repository.save(show)

        assertSoftly {
            it.assertThat(show.id).isEqualTo(ShowUtils.SHOWS_COUNT + 1)
            it.assertThat(show.createdUser).isEqualTo(TestConstants.ACCOUNT.uuid!!)
            it.assertThat(show.createdTime).isEqualTo(TestConstants.TIME)
            it.assertThat(show.updatedUser).isEqualTo(TestConstants.ACCOUNT.uuid!!)
            it.assertThat(show.updatedTime).isEqualTo(TestConstants.TIME)
        }

        val addedShow = ShowUtils.getShow(entityManager, ShowUtils.SHOWS_COUNT + 1)!!
        assertThat(addedShow).isNotNull
        ShowUtils.assertShowDeepEquals(expected = expectedShow, actual = addedShow)

        assertSoftly {
            it.assertThat(ShowUtils.getShowsCount(entityManager = entityManager)).isEqualTo(ShowUtils.SHOWS_COUNT + 1)
            it.assertThat(SeasonUtils.getSeasonsCount(entityManager = entityManager)).isEqualTo(SeasonUtils.SEASONS_COUNT)
            it.assertThat(EpisodeUtils.getEpisodesCount(entityManager = entityManager)).isEqualTo(EpisodeUtils.EPISODES_COUNT)
        }
    }

    /**
     * Test method for update show.
     */
    @Test
    fun update() {
        val show = ShowUtils.updateShow(entityManager = entityManager, id = 1)
        val expectedShow = ShowUtils.getShowDomain(index = 1)
            .updated()
            .copy(position = ShowUtils.POSITION)
            .fillAudit(audit = AuditUtils.updatedAudit())

        repository.saveAndFlush(show)

        val updatedShow = ShowUtils.getShow(entityManager = entityManager, id = 1)
        assertThat(updatedShow).isNotNull
        ShowUtils.assertShowDeepEquals(expected = expectedShow, actual = updatedShow!!)

        assertSoftly {
            it.assertThat(ShowUtils.getShowsCount(entityManager = entityManager)).isEqualTo(ShowUtils.SHOWS_COUNT)
            it.assertThat(SeasonUtils.getSeasonsCount(entityManager = entityManager)).isEqualTo(SeasonUtils.SEASONS_COUNT)
            it.assertThat(EpisodeUtils.getEpisodesCount(entityManager = entityManager)).isEqualTo(EpisodeUtils.EPISODES_COUNT)
        }
    }

    /**
     * Test method for remove show.
     */
    @Test
    fun remove() {
        repository.delete(ShowUtils.getShow(entityManager = entityManager, id = 1)!!)

        assertThat(ShowUtils.getShow(entityManager = entityManager, id = 1)).isNull()

        assertSoftly {
            it.assertThat(ShowUtils.getShowsCount(entityManager = entityManager)).isEqualTo(ShowUtils.SHOWS_COUNT - 1)
            it.assertThat(SeasonUtils.getSeasonsCount(entityManager = entityManager)).isEqualTo(SeasonUtils.SEASONS_COUNT - SeasonUtils.SEASONS_PER_SHOW_COUNT)
            it.assertThat(EpisodeUtils.getEpisodesCount(entityManager = entityManager)).isEqualTo(EpisodeUtils.EPISODES_COUNT - EpisodeUtils.EPISODES_PER_SHOW_COUNT)
        }
    }

    /**
     * Test method for remove all shows.
     */
    @Test
    fun removeAll() {
        repository.deleteAll()

        assertSoftly {
            it.assertThat(ShowUtils.getShowsCount(entityManager = entityManager)).isEqualTo(0)
            it.assertThat(SeasonUtils.getSeasonsCount(entityManager = entityManager)).isEqualTo(0)
            it.assertThat(EpisodeUtils.getEpisodesCount(entityManager = entityManager)).isEqualTo(0)
        }
    }

    /**
     * Test method for get shows for user.
     */
    @Test
    fun findByCreatedUser() {
        val shows = repository.findByCreatedUser(user = AuditUtils.getAudit().createdUser!!)

        ShowUtils.assertDomainShowsDeepEquals(expected = ShowUtils.getShows(), actual = shows)

        assertSoftly {
            it.assertThat(ShowUtils.getShowsCount(entityManager = entityManager)).isEqualTo(ShowUtils.SHOWS_COUNT)
            it.assertThat(SeasonUtils.getSeasonsCount(entityManager = entityManager)).isEqualTo(SeasonUtils.SEASONS_COUNT)
            it.assertThat(EpisodeUtils.getEpisodesCount(entityManager = entityManager)).isEqualTo(EpisodeUtils.EPISODES_COUNT)
        }
    }

    /**
     * Test method for get show by id for user.
     */
    @Test
    fun findByIdAndCreatedUser() {
        val user = AuditUtils.getAudit().createdUser!!
        for (i in 1..ShowUtils.SHOWS_COUNT) {
            val author = repository.findByIdAndCreatedUser(id = i, user = user).orElse(null)

            ShowUtils.assertShowDeepEquals(expected = ShowUtils.getShowDomain(index = i), actual = author)
        }

        assertThat(repository.findByIdAndCreatedUser(id = Int.MAX_VALUE, user = user)).isNotPresent

        assertSoftly {
            it.assertThat(ShowUtils.getShowsCount(entityManager = entityManager)).isEqualTo(ShowUtils.SHOWS_COUNT)
            it.assertThat(SeasonUtils.getSeasonsCount(entityManager = entityManager)).isEqualTo(SeasonUtils.SEASONS_COUNT)
            it.assertThat(EpisodeUtils.getEpisodesCount(entityManager = entityManager)).isEqualTo(EpisodeUtils.EPISODES_COUNT)
        }
    }

}
